urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/api/LauncherAPI.java

```diff
@@ -0,0 +1,335 @@
+    {
+        return postUrl(dest, params, null);
+    }
+    
+    public String postUrl(String dest, Map<String, String> params,
+            String keyFile)
+    {
+        String result = null;
+        HttpURLConnection connection = null;
+        
+        try
+        {
+            final URL url = new URL(dest);
+            connection = (HttpURLConnection) url.openConnection();
+            
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type",
+                    "application/x-www-form-urlencoded");
+            
+            final StringBuffer sb = new StringBuffer();
+            final Iterator<Entry<String, String>> iterator = params.entrySet()
+                    .iterator();
+            while (iterator.hasNext())
+            {
+                final Entry<String, String> param = iterator.next();
+                
+                sb.append(param.getKey());
+                sb.append('=');
+                sb.append(param.getValue());
+                
+                if (iterator.hasNext())
+                {
+                    sb.append('&');
+                }
+            }
+            
+            connection.setRequestProperty("Content-Length",
+                    Integer.toString(sb.toString().getBytes().length));
+            connection.setRequestProperty("Content-Language", "en-US");
+            
+            connection.setUseCaches(false);
+            connection.setDoInput(true);
+            connection.setDoOutput(true);
+            
+            connection.connect();
+            
+            if (url.getProtocol().contains("https") && keyFile != null)
+            {
+                final Certificate[] certs = ((HttpsURLConnection) connection)
+                        .getServerCertificates();
+                
+                final byte[] bytes = new byte[294];
+                final DataInputStream dis = new DataInputStream(
+                        LauncherAPI.class
+                                .getResourceAsStream("keys/" + keyFile));
+                dis.readFully(bytes);
+                dis.close();
+                
+                final Certificate c = certs[0];
+                final PublicKey pk = c.getPublicKey();
+                final byte[] data = pk.getEncoded();
+                
+                for (int i = 0; i < data.length; i++)
+                {
+                    if (data[i] == bytes[i])
+                    {
+                        continue;
+                    }
+                    throw new RuntimeException("Public key mismatch");
+                }
+            }
+            
+            final DataOutputStream out = new DataOutputStream(
+                    connection.getOutputStream());
+            out.writeBytes(sb.toString());
+            out.flush();
+            out.close();
+            
+            final InputStream in = connection.getInputStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Use a standard utility function such as Apache Commons IO `IOUtils.toString(InputStream, Charset)` to read data from an input stream. Using a standard utility function makes your code easier to read and less prone to resource leaks. Do not write your own code to read data from the input stream. See the following diff on GitHub that makes this code change: [diff1](https://github.com/MyRobotLab/myrobotlab/commit/a4db462136ebf6c9cb4bc1e60bfac6d6f37a4b6e#diff-a9871318c9e95407fbecd277bb0f4729724c94216d5d5d57d13f743058b9c76dL242)

Learn more: [IOUtils.toString in the Apache Commons IO 2.11.0 API](https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#toString-java.io.InputStream-java.nio.charset.Charset-)